crates/bevy_ecs/src/storage/resource_non_send.rs

```diff
@@ -0,0 +1,522 @@
+
+            result
+        }
+
+        fn run_indirect<F, T>(&mut self, f: F) -> T
+        where
+            F: FnOnce(&mut ThreadLocals) -> T + Send,
+            T: Send + 'static,
+        {
+            let ThreadLocalAccess::Indirect(ref mut sender) = self.access else { unreachable!() };
+
+            let (result_tx, result_rx) = std::sync::mpsc::sync_channel(1);
+            let f = |tls: &mut ThreadLocals| {
+                tls.update_change_tick();
+                let saved = std::mem::replace(&mut tls.last_tick, *self.last_run);
+                // we want to propagate to caller instead of panicking in the main thread
+                let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| f(tls)));
+                tls.last_tick = saved;
+
+                *self.last_run = tls.last_tick;
+                result_tx.send(result).unwrap();
+            };
+
+            let f: Box<dyn FnOnce(&mut ThreadLocals) + Send> = Box::new(f);
+
+            // SAFETY: `scope` blocks the calling thread until `f` completes,
+            // so any borrows within `f` will remain valid until then.
+            let f: Box<dyn FnOnce(&mut ThreadLocals) + Send + 'static> =
+                unsafe { std::mem::transmute(f) };
+
+            // Send task to the main thread.
+            sender
+                .send_task(f)
+                .unwrap_or_else(|_| panic!("receiver missing"));
+
+            // Wait to receive result back from the main thread.
+            result_rx.recv().unwrap().unwrap()
+        }
+    }
+
+    // SAFETY: This system param does not borrow any data from the world.
+    unsafe impl SystemParam for ThreadLocal<'_, '_> {
+        type State = ThreadLocalState;
+        type Item<'w, 's> = ThreadLocal<'w, 's>;
+
+        fn init_state(
+            world: &mut crate::prelude::World,
+            system_meta: &mut crate::system::SystemMeta,
+        ) -> Self::State {
+            let component_id =
+                crate::system::ResMut::<ThreadLocalAccessor>::init_state(world, system_meta);
+
+            ThreadLocalState {
+                component_id,
+                last_run: Tick::new(0),
+            }
+        }
+
+        unsafe fn get_param<'world, 'state>(
+            state: &'state mut Self::State,
+            system_meta: &crate::system::SystemMeta,
+            world: UnsafeWorldCell<'world>,
+            curr_tick: Tick,
+        ) -> Self::Item<'world, 'state> {
+            let mut accessor = crate::system::ResMut::<ThreadLocalAccessor>::get_param(
+                &mut state.component_id,
+                system_meta,
+                world,
+                curr_tick,
+            );
+
+            let access = if std::thread::current().id() == accessor.owning_thread {
+                // SAFETY: We are on the owning thread and pointer is valid.
+                let ptr = accessor.direct;
+                unsafe { ThreadLocalAccess::Direct(&mut *ptr) }
```

Comment:
Also, are we sure there's no other direct access to the thread local? For example, couldn't we access it through the `ThreadLocalAccess` that exist in a `SubApp`? In that case we would have two mutable references to the same shared thread local storage, which would be very bad.

Reply:
No, unless it's somehow possible for two systems to run on the same thread at the same time. There's also no public API on `ThreadLocalAccessor` that lets you read the pointer.

Reply:
Is it? I'm not familiar with what happens to references when a future is awaited and yields to another, but `ThreadLocals` is non-`Send` and non-`Sync`, so it can only be referenced on the same thread, so AFAIK it's impossible to see aliasing.

Reply:
I was thinking of something like this:

```rs
fn bad_subapp_extract(main_world: &mut World, subapp: &mut App) {
    let main_tl_state = SystemState::<ThreadLocal>::new(main_world);
    let main_tl = main_tl_state.get_mut(main_world);

    // This may run on this thread or another thread, doesn't really matter.
    main_tl.run(|tls1| {
        // Here we're in the tls thread, so this will get direct access to the tls
        let subapp_tl_state = SystemState::<ThreadLocal>::new(&mut subapp.world);
        let subapp_tl = subapp_tl_state.get_mut(&mut subapp.world);

        // Thus this should run immediately in this thread
        subapp_tl.run(|tls2| {
            // Now we have access to both tls1 and tls2, bad!
        });
    });
}
```

Where would this fail?

Reply:
:/ Well, I stand corrected.

Could either:
- Get rid of the direct access and only have it when the target is `wasm32` (or some compiler flag is enabled).
- Work a reference counter in there somehow and panic if there is more than one reference.

Reply:
OK, the current solution I've come up with is to use a wrapped `Box<Mutex<ThreadLocals>` so that the pointer stays valid and every borrow has to acquire the lock to create the `&mut ThreadLocals`.